samza-core/src/main/java/org/apache/samza/zk/ScheduleAfterDebounceTime.java

```diff
@@ -20,99 +20,156 @@
         } catch (Exception e) {
           // we ignore the exception
-          LOG.warn("cancel for action " + actionName + " failed with ", e);
+          LOG.warn("Cancelling the future of action: {} failed.", actionName, e);
         }
       }
       futureHandles.remove(actionName);
     }
-    // schedule a new task
-    sf = scheduledExecutorService.schedule(() -> {
-        try {
-          runnable.run();
-          LOG.debug(actionName + " completed successfully.");
-        } catch (Throwable t) {
-          LOG.error(actionName + " threw an exception.", t);
-          if (scheduledTaskFailureCallback != null) {
-            scheduledTaskFailureCallback.onError(t);
-          }
+  }
+
+  /**
+   * Decorate the executable action with exception handlers to facilitate cleanup on failures.
+   *
+   * @param actionName the name of the scheduleable action.
+   * @param runnable the action to execute.
+   * @return the executable action decorated with exception handlers.
+   */
+  private Runnable getScheduleableAction(String actionName, Runnable runnable) {
+    return () -> {
+      try {
+        runnable.run();
+        /*
+         * Expects all run() implementations <b>not to swallow the interrupts.</b>
+         * This thread is interrupted from an external source(mostly executor service) to die.
+         */
+        if (Thread.currentThread().isInterrupted()) {
+          LOG.warn("Action: {} is interrupted.", actionName);
+          doCleanUpOnTaskException(new InterruptedException());
+        } else {
+          LOG.debug("Action: {} completed successfully.", actionName);
         }
-      },
-     debounceTimeMs,
-     TimeUnit.MILLISECONDS);
-    LOG.info("scheduled " + actionName + " in " + debounceTimeMs);
-    futureHandles.put(actionName, sf);
+      } catch (Exception exception) {
+        LOG.error("Execution of action: {} failed.", actionName, exception);
+        doCleanUpOnTaskException(exception);
+      }
+    };
   }
 
-  public void stopScheduler() {
-    // shutdown executor service
-    scheduledExecutorService.shutdown();
+  /**
+   * Handler method to invoke on an exception during an scheduled task execution and does
+   * the following operations in sequential order.
+   * <ul>
+   *   <li> Stops the scheduler.</li>
+   *   <li> Invokes the onError handler method if taskCallback is defined.</li>
+   * </ul>
+   *
+   * @param exception the exception happened during task execution.
+   */
+  private void doCleanUpOnTaskException(Exception exception) {
+    // 1. Stop the scheduler. If a task execution fails or a task is interrupted, scheduler will not accept/execute any new tasks.
+    stopScheduler();
+
+    // 2. Invoke the callback if its defined.
+    scheduledTaskCallback.ifPresent(callback -> callback.onError(exception));
   }
 
-  interface ScheduledTaskFailureCallback {
+  /**
+   * A ScheduledTaskCallback is invoked on first occurrence of exception
```

Comment:
It is either generic callback interface or error specific one, so either rename the interface back to ScheduledTaskFailureCallback or in the comment write ScheduledTaskCallback::onError().

Reply:
Done. Updated the comment.

Reply:
@navina 
Myself & boris discussed offline, there're two approaches to not take in ScheduledTaskCallbak as constructor param (It looks kind of weird to take in callback as param as constructor param).
i) have setScheduledTaskCallback method here in this class
ii) Or taken in scheduleTaskCallBack as a param to schedule method itself. 

ii) involves touching/modifying in lot of places, hence going with i) which is reasonable as well. Do you agree with ii)

Reply:
@shanthoosh thanks for listing the approaches. I agree that ii) is the ideal solution. For now, it's fine to keep it in setScheduledTaskCallback.